velero-plugins/build/restore.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/heptio/velero/pkg/apis/velero/v1"
+	"github.com/heptio/velero/pkg/restore"
+	buildv1API "github.com/openshift/api/build/v1"
+	"github.com/sirupsen/logrus"
+	corev1API "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"strings"
+)
+
+// MyRestorePlugin is a restore item action plugin for Velero
+type RestorePlugin struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns a restore.ResourceSelector that applies to everything
+func (p *RestorePlugin) AppliesTo() (restore.ResourceSelector, error) {
+	return restore.ResourceSelector{
+		IncludedResources: []string{"builds"},
+	}, nil
+}
+
+func (p *RestorePlugin) Execute(item runtime.Unstructured, restore *v1.Restore) (runtime.Unstructured, error, error) {
+	p.Log.Info("Hello from Build RestorePlugin!")
+
+	metadata, err := meta.Accessor(item)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	annotations := metadata.GetAnnotations()
```

Comment:
Since we're pulling the full typed resource below with `build := buildv1API.Build`, it's probably cleaner to use this to grab the annotation as well, like we're doing with the imagestream backup plugin. Also, I found with the imagestream plugin that this method of accessing the annotations didn't work well on the save end in conjunction with the json.Unmarshall code -- the annotations weren't being set.

`annotations := build.Annotations` will grab the annotations, and once you're done modifying the returned resource, `build.Annotations = annotations` would replace the `metadata.SetAnnotations()` call at the end.

Reply:
Good catch, thanks! I don't know that I even need to be updating the annotations I just have been copying the format of the other plugins. I can just remove it unless you think we should always annotate which plugin it hits.

Reply:
I don't have a strong opinion on the "always annotate which plugin it hits" question, but if you keep it, it's probably better to update it to use the Build object.

Reply:
I just removed the annotation since it wasn't being used for anything